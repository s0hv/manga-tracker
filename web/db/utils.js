import {
  NUMERIC_VALUE_OUT_OF_RANGE,
  INVALID_TEXT_REPRESENTATION,
  UNIQUE_VIOLATION,
  IN_FAILED_SQL_TRANSACTION,
  FOREIGN_KEY_VIOLATION,
  NOT_NULL_VIOLATION,
} from 'pg-error-constants';

import { pgp } from '.';
import { NoColumnsError } from './errors.js';
import { dbLogger } from '../utils/logging.js';
import { StatusError } from '../utils/errors.js';

/**
 * Generate update statement from an object while filtering out undefined values.
 * The generated update statement lacks a WHERE clause and has the values embedded in the query.
 * @param {Object} o Input object
 * @param {String} tableName Name of the table to be updated
 * @returns {String} The update statement without a whereclause
 */
export const generateUpdate = (o, tableName) => {
  const obj = { ...o };
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key]);
  try {
    return pgp.helpers.update(obj, null, tableName);
  } catch (e) {
    throw new NoColumnsError('No valid columns given');
  }
};

export function handleError(err, res, msgOverrides = {}) {
  if (typeof err?.getErrors === 'function') {
    err = err.getErrors().filter(e => e?.code !== IN_FAILED_SQL_TRANSACTION)[0] || err;
  }

  if (err instanceof StatusError) {
    res.status(err.status).json({ error: err.message });
    return;
  }

  if (err instanceof NoColumnsError) {
    res.status(400).json({ error: err.message || 'No valid columns given' });
    return;
  }

  const msg = msgOverrides[err.code];
  if (err.code === INVALID_TEXT_REPRESENTATION) {
    dbLogger.debug(err.message);
    res.status(400).json({ error: msg || 'Invalid data type given' });
  } else if (err.code === NUMERIC_VALUE_OUT_OF_RANGE) {
    res.status(400).json({ error: msg || 'Number value out of range' });
  } else if (err.code === UNIQUE_VIOLATION) {
    res.status(422).json({ error: msg || 'Resource already exists' });
  } else if (err.code === FOREIGN_KEY_VIOLATION) {
    res.status(404).json({ error: msg || 'Foreign key violation' });
  } else if (err.code === NOT_NULL_VIOLATION) {
    res.status(400).json({ error: msg || 'Not null value was null' });
  } else {
    dbLogger.error(err, 'Unknown database error');
    res.status(500).json({ error: msg || 'Internal server error' });
  }
}
