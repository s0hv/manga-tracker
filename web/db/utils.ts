import {
  FOREIGN_KEY_VIOLATION,
  IN_FAILED_SQL_TRANSACTION,
  INVALID_TEXT_REPRESENTATION,
  NOT_NULL_VIOLATION,
  NUMERIC_VALUE_OUT_OF_RANGE,
  UNIQUE_VIOLATION,
} from 'pg-error-constants';
import type { Response } from 'express-serve-static-core';

import { NoColumnsError } from './errors';
import { dbLogger } from '../utils/logging.js';
import { StatusError } from '../utils/errors.js';
import type { Db } from '.';

/**
 * Generate update statement from an object while filtering out undefined values.
 * Do not pass untrusted properties to this method, as it will update every column given to it
 * @param {Object} o Input object
 * @param {Db} sql Database instance
 */
export const generateUpdate = (o: any, sql: Db) => {
  const obj = { ...o };
  Object.keys(obj).forEach(key => obj[key] === undefined && delete obj[key]);

  if (Object.keys(obj).length === 0) throw new NoColumnsError('No valid columns given');

  return sql(obj);
};

export function handleError(err: any, res: Response, msgOverrides: any = {}) {
  if (typeof err?.getErrors === 'function') {
    err = err.getErrors().filter((e: any) => e?.code !== IN_FAILED_SQL_TRANSACTION)[0] || err;
  }

  if (err instanceof StatusError) {
    res.status(err.status).json({ error: err.message });
    return;
  }

  if (err instanceof NoColumnsError) {
    res.status(400).json({ error: err.message || 'No valid columns given' });
    return;
  }

  const msg = msgOverrides[err.code];
  if (err.code === INVALID_TEXT_REPRESENTATION) {
    dbLogger.debug(err.message);
    res.status(400).json({ error: msg || 'Invalid data type given' });
  } else if (err.code === NUMERIC_VALUE_OUT_OF_RANGE) {
    res.status(400).json({ error: msg || 'Number value out of range' });
  } else if (err.code === UNIQUE_VIOLATION) {
    res.status(422).json({ error: msg || 'Resource already exists' });
  } else if (err.code === FOREIGN_KEY_VIOLATION) {
    res.status(404).json({ error: msg || 'Foreign key violation' });
  } else if (err.code === NOT_NULL_VIOLATION) {
    res.status(400).json({ error: msg || 'Not null value was null' });
  } else {
    dbLogger.error(err, 'Unknown database error');
    res.status(500).json({ error: msg || 'Internal server error' });
  }
}
