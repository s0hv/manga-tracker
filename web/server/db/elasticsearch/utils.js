export const handleElasticError = (err, res) => {
  const error = 'Internal server error';
  console.error(err);
  res.status(500).json({ error });
};

export const extractFields = (result, fields, idPrefix, customFieldFormatter) => {
  idPrefix = idPrefix ? `${idPrefix}_` : '';
  return result.body.hits.hits.map(hit => {
    const o = {
      [`${idPrefix}id`]: Number(hit._id),
      score: hit._score,
    };
    if (fields) {
      fields.forEach(field => {
        o[field] = (hit.fields[field] || [])[0];
      });

      if (customFieldFormatter) {
        const formatted = customFieldFormatter(hit.fields);
        Object.keys(formatted).forEach(k => { o[k] = formatted[k] });
      }
    }

    return o;
  });
};
