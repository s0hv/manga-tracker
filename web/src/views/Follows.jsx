import React from 'react';

import { Container, IconButton, Paper, Typography, } from '@mui/material';
import { lighten, styled } from '@mui/material/styles';
import { RssFeed as RssFeedIcon } from '@mui/icons-material';
import PropTypes from 'prop-types';

import FollowsComponent from '../components/Follows';
import { useUser } from '../utils/useUser';

const FollowCard = styled(Paper)(({ theme }) => ({
  height: '100%',
  overflow: 'hidden',
  marginBottom: theme.spacing(2),
  backgroundColor: lighten(theme.palette.background.paper, 0.05),
}));

const TopRow = styled('div')({
  display: 'flex',
  justifyContent: 'space-between',
});

const Follows = (props) => {
  const { user } = useUser();

  return (
    <Container maxWidth='lg'>
      <FollowCard>
        <TopRow>
          <Typography variant='h2' sx={{ ml: 2, mt: 2 }}>
            Follows
          </Typography>
          <IconButton
            href={`/rss/${user.uuid.replace(/-/g, '')}`}
            target='_blank'
            sx={{ alignSelf: 'center', height: '100%' }}
            size='medium'
            aria-label='Follows RSS feed'
          >
            <RssFeedIcon fontSize='large' />
          </IconButton>
        </TopRow>
        <FollowsComponent {...props} />
      </FollowCard>
    </Container>
  );
};

Follows.propTypes = {
  follows: PropTypes.arrayOf(PropTypes.object),
};

export default Follows;
