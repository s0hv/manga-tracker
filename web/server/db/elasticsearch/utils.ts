import { ApiResponse } from '@elastic/elasticsearch';
import type { Response } from 'express-serve-static-core';

import type {
  MangaSearchResult,
  MangaSearchResultFields,
} from '@/db/elasticsearch/manga';

export const handleElasticError = (err: unknown, res: Response) => {
  const error = 'Internal server error';
  console.error(err);
  res.status(500).json({ error });
};

export type ExtractedFields = {
  score: number
  [key: string]: unknown
};
export type CustomFieldFormatter<T extends boolean> = (fields: MangaSearchResultFields<T>) => Record<string, unknown>;

export const extractFields = <T extends boolean>(
  result: ApiResponse<MangaSearchResult<T>>,
  fields?: (keyof MangaSearchResultFields<T>)[],
  idPrefix?: string,
  customFieldFormatter?: (fields: MangaSearchResultFields<T>) => Record<string, unknown>
) => {
  const mappedIdPrefix = idPrefix ? `${idPrefix}_` : '';
  return result.body.hits.hits.map(hit => {
    const o: ExtractedFields = {
      [`${mappedIdPrefix}id`]: Number(hit._id),
      score: hit._score,
    };
    if (fields) {
      fields.forEach(field => {
        o[field as string] = (hit.fields[field] as Array<unknown> ?? [])[0];
      });

      if (customFieldFormatter) {
        const formatted = customFieldFormatter(hit.fields);
        Object.keys(formatted).forEach(k => {
          o[k] = formatted[k];
        });
      }
    }

    return o;
  });
};
