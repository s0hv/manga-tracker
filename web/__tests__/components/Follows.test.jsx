import React from 'react';
import { render, screen, within } from '@testing-library/react';
import Follows from '../../src/components/Follows';
import { mockUTCDates } from '../utils';

mockUTCDates();

describe('Follows page should render correctly', () => {
  const follows = [
    {
      title: 'Dr. STONE',
      cover: '/images/test.png',
      mangaId: 1,
      latestChapter: 90,
      latestRelease: '2020-08-20T15:36:07.865Z',
      services: [
        {
          serviceId: 1,
          serviceName: 'MANGA Plus',
          titleId: '100010',
          url: 'https://mangaplus.shueisha.co.jp/titles/{}',
        },
        {
          serviceId: 2,
          serviceName: 'MangaDex',
          titleId: '20882',
          url: 'https://mangadex.org/title/{}',
        },
      ],
      followedServices: [
        1,
        null,
      ],
    },
    {
      title: 'Shuumatsu no Valkyrie',
      cover: '/images/test2.png',
      mangaId: 8,
      services: [
        {
          serviceId: 2,
          serviceName: 'MangaDex',
          titleId: '33537',
          url: 'https://mangadex.org/title/{}',
        },
      ],
      followedServices: [
        null,
      ],
    },
    {
      title: 'Kengan Omega',
      cover: '/images/test3.png',
      mangaId: 64,
      latestChapter: 10,
      services: [
        {
          serviceId: 2,
          serviceName: 'MangaDex',
          titleId: '33538',
          url: 'https://mangadex.org/title/{}',
        },
      ],
      followedServices: [
        null,
      ],
    },
  ];

  it('should render no text without follows', () => {
    render(<Follows />);
    expect(screen.queryByText(/.+/)).not.toBeInTheDocument();
  });

  it('should render correctly with follows', () => {
    render(<Follows follows={follows} />);

    expect(follows.length).toBeGreaterThan(0);

    follows.forEach(follow => {
      const elem = within(screen.getByText(follow.title).parentElement);

      if (!follow.latestRelease) {
        expect(elem.getByRole('row', { name: /latest release: unknown/i })).toBeInTheDocument();
      } else {
        expect(elem.getByRole('row', { name: /latest release: \d+ \w+ ago/i })).toBeInTheDocument();
      }

      if (!follow.latestChapter) {
        expect(elem.getByRole('row', { name: /latest chapter: no chapters/i })).toBeInTheDocument();
      } else {
        expect(elem.getByRole('row', { name: /latest chapter: \d+/i })).toBeInTheDocument();
      }

      const serviceList = within(elem.getByLabelText('manga services'));
      expect(serviceList.getByText(/all services/i)).toBeInTheDocument();
      follow.services.forEach(service => {
        expect(serviceList.getByText(service.serviceName)).toBeInTheDocument();
      });

      expect(serviceList.getAllByRole('button', { name: /(un)?follow/i })).toHaveLength(follow.services.length+1);
    });
  });
});
