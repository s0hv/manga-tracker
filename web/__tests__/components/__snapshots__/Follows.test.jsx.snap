// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Follows page should render correctly should render correctly with follows 1`] = `
<div
  className="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthLg"
>
  <div
    className="MuiPaper-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
      >
        <h4
          className="MuiTypography-root makeStyles-followTitle-2 MuiTypography-h4 MuiTypography-alignCenter"
        >
          Dr. STONE
        </h4>
        <div
          className="makeStyles-followContent-3"
        >
          <div
            className="makeStyles-thumbnail-5"
          >
            <a
              href="/manga/1"
              onClick={[Function]}
              onMouseEnter={[Function]}
              target="_blank"
            >
              <div
                style={
                  Object {
                    "bottom": 0,
                    "boxSizing": "border-box",
                    "display": "block",
                    "left": 0,
                    "margin": 0,
                    "overflow": "hidden",
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <noscript>
                  <img
                    alt="Dr. STONE"
                    decoding="async"
                    sizes="(max-width: 600px) 192px, 256px"
                    src="/_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=3840&q=75"
                    srcSet="/_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=16&q=75 16w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=32&q=75 32w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=48&q=75 48w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=64&q=75 64w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=96&q=75 96w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=128&q=75 128w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=256&q=75 256w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=384&q=75 384w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=640&q=75 640w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=750&q=75 750w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=828&q=75 828w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimages%2Ftest.png.256.jpg&w=3840&q=75 3840w"
                    style={
                      Object {
                        "border": "none",
                        "bottom": 0,
                        "boxSizing": "border-box",
                        "display": "block",
                        "height": 0,
                        "left": 0,
                        "margin": "auto",
                        "maxHeight": "100%",
                        "maxWidth": "100%",
                        "minHeight": "100%",
                        "minWidth": "100%",
                        "objectFit": "contain",
                        "objectPosition": undefined,
                        "padding": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                        "width": 0,
                      }
                    }
                  />
                </noscript>
                <img
                  alt="Dr. STONE"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style={
                    Object {
                      "border": "none",
                      "bottom": 0,
                      "boxSizing": "border-box",
                      "display": "block",
                      "height": 0,
                      "left": 0,
                      "margin": "auto",
                      "maxHeight": "100%",
                      "maxWidth": "100%",
                      "minHeight": "100%",
                      "minWidth": "100%",
                      "objectFit": "contain",
                      "objectPosition": undefined,
                      "padding": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                      "width": 0,
                    }
                  }
                />
              </div>
            </a>
          </div>
          <div
            className="makeStyles-followDetails-4"
          >
            <table>
              <tbody>
                <tr>
                  <td>
                    <p
                      className="MuiTypography-root MuiTypography-body1"
                    >
                      Latest release: 
                    </p>
                  </td>
                  <td>
                    <p
                      className="MuiTypography-root MuiTypography-body1"
                    >
                      1 day ago
                    </p>
                  </td>
                </tr>
                <tr>
                  <td>
                    <p
                      className="MuiTypography-root MuiTypography-body1"
                    >
                      Latest chapter: 
                    </p>
                  </td>
                  <td>
                    <p
                      className="MuiTypography-root MuiTypography-body1"
                    >
                      90
                    </p>
                  </td>
                </tr>
              </tbody>
            </table>
            <ul
              className="MuiList-root makeStyles-serviceList-6 MuiList-padding"
            >
              <li
                className="MuiListItem-root makeStyles-followService-7"
                disabled={false}
              >
                <div
                  className="MuiListItemText-root makeStyles-serviceName-8"
                >
                  <span
                    className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    All services
                  </span>
                </div>
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    Unfollow
                  </span>
                </button>
              </li>
              <li
                className="MuiListItem-root makeStyles-followService-7"
                disabled={false}
              >
                <div
                  className="MuiListItemText-root makeStyles-serviceName-8"
                >
                  <span
                    className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    MANGA Plus
                  </span>
                </div>
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    Unfollow
                  </span>
                </button>
              </li>
              <li
                className="MuiListItem-root makeStyles-followService-7"
                disabled={false}
              >
                <div
                  className="MuiListItemText-root makeStyles-serviceName-8"
                >
                  <span
                    className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    MangaDex
                  </span>
                </div>
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    Follow
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
      >
        <h4
          className="MuiTypography-root makeStyles-followTitle-2 MuiTypography-h4 MuiTypography-alignCenter"
        >
          Shuumatsu no Valkyrie
        </h4>
        <div
          className="makeStyles-followContent-3"
        >
          <div
            className="makeStyles-thumbnail-5"
          >
            <a
              href="/manga/8"
              onClick={[Function]}
              onMouseEnter={[Function]}
              target="_blank"
            >
              <div
                style={
                  Object {
                    "bottom": 0,
                    "boxSizing": "border-box",
                    "display": "block",
                    "left": 0,
                    "margin": 0,
                    "overflow": "hidden",
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <noscript>
                  <img
                    alt="Shuumatsu no Valkyrie"
                    decoding="async"
                    sizes="(max-width: 600px) 192px, 256px"
                    src="/_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=3840&q=75"
                    srcSet="/_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=16&q=75 16w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=32&q=75 32w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=48&q=75 48w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=64&q=75 64w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=96&q=75 96w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=128&q=75 128w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=256&q=75 256w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=384&q=75 384w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=640&q=75 640w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=750&q=75 750w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=828&q=75 828w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimages%2Ftest2.png.256.jpg&w=3840&q=75 3840w"
                    style={
                      Object {
                        "border": "none",
                        "bottom": 0,
                        "boxSizing": "border-box",
                        "display": "block",
                        "height": 0,
                        "left": 0,
                        "margin": "auto",
                        "maxHeight": "100%",
                        "maxWidth": "100%",
                        "minHeight": "100%",
                        "minWidth": "100%",
                        "objectFit": "contain",
                        "objectPosition": undefined,
                        "padding": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                        "width": 0,
                      }
                    }
                  />
                </noscript>
                <img
                  alt="Shuumatsu no Valkyrie"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style={
                    Object {
                      "border": "none",
                      "bottom": 0,
                      "boxSizing": "border-box",
                      "display": "block",
                      "height": 0,
                      "left": 0,
                      "margin": "auto",
                      "maxHeight": "100%",
                      "maxWidth": "100%",
                      "minHeight": "100%",
                      "minWidth": "100%",
                      "objectFit": "contain",
                      "objectPosition": undefined,
                      "padding": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                      "width": 0,
                    }
                  }
                />
              </div>
            </a>
          </div>
          <div
            className="makeStyles-followDetails-4"
          >
            <table>
              <tbody>
                <tr>
                  <td>
                    <p
                      className="MuiTypography-root MuiTypography-body1"
                    >
                      Latest release: 
                    </p>
                  </td>
                  <td>
                    <p
                      className="MuiTypography-root MuiTypography-body1"
                    >
                      Unknown
                    </p>
                  </td>
                </tr>
                <tr>
                  <td>
                    <p
                      className="MuiTypography-root MuiTypography-body1"
                    >
                      Latest chapter: 
                    </p>
                  </td>
                  <td>
                    <p
                      className="MuiTypography-root MuiTypography-body1"
                    >
                      No chapters
                    </p>
                  </td>
                </tr>
              </tbody>
            </table>
            <ul
              className="MuiList-root makeStyles-serviceList-6 MuiList-padding"
            >
              <li
                className="MuiListItem-root makeStyles-followService-7"
                disabled={false}
              >
                <div
                  className="MuiListItemText-root makeStyles-serviceName-8"
                >
                  <span
                    className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    All services
                  </span>
                </div>
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    Unfollow
                  </span>
                </button>
              </li>
              <li
                className="MuiListItem-root makeStyles-followService-7"
                disabled={false}
              >
                <div
                  className="MuiListItemText-root makeStyles-serviceName-8"
                >
                  <span
                    className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    MangaDex
                  </span>
                </div>
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    Follow
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
      >
        <h4
          className="MuiTypography-root makeStyles-followTitle-2 MuiTypography-h4 MuiTypography-alignCenter"
        >
          Kengan Omega
        </h4>
        <div
          className="makeStyles-followContent-3"
        >
          <div
            className="makeStyles-thumbnail-5"
          >
            <a
              href="/manga/64"
              onClick={[Function]}
              onMouseEnter={[Function]}
              target="_blank"
            >
              <div
                style={
                  Object {
                    "bottom": 0,
                    "boxSizing": "border-box",
                    "display": "block",
                    "left": 0,
                    "margin": 0,
                    "overflow": "hidden",
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <noscript>
                  <img
                    alt="Kengan Omega"
                    decoding="async"
                    sizes="(max-width: 600px) 192px, 256px"
                    src="/_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=3840&q=75"
                    srcSet="/_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=16&q=75 16w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=32&q=75 32w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=48&q=75 48w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=64&q=75 64w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=96&q=75 96w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=128&q=75 128w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=256&q=75 256w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=384&q=75 384w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=640&q=75 640w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=750&q=75 750w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=828&q=75 828w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=1080&q=75 1080w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=1200&q=75 1200w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=1920&q=75 1920w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=2048&q=75 2048w, /_next/image?url=%2Fimages%2Ftest3.png.256.jpg&w=3840&q=75 3840w"
                    style={
                      Object {
                        "border": "none",
                        "bottom": 0,
                        "boxSizing": "border-box",
                        "display": "block",
                        "height": 0,
                        "left": 0,
                        "margin": "auto",
                        "maxHeight": "100%",
                        "maxWidth": "100%",
                        "minHeight": "100%",
                        "minWidth": "100%",
                        "objectFit": "contain",
                        "objectPosition": undefined,
                        "padding": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                        "width": 0,
                      }
                    }
                  />
                </noscript>
                <img
                  alt="Kengan Omega"
                  decoding="async"
                  src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                  style={
                    Object {
                      "border": "none",
                      "bottom": 0,
                      "boxSizing": "border-box",
                      "display": "block",
                      "height": 0,
                      "left": 0,
                      "margin": "auto",
                      "maxHeight": "100%",
                      "maxWidth": "100%",
                      "minHeight": "100%",
                      "minWidth": "100%",
                      "objectFit": "contain",
                      "objectPosition": undefined,
                      "padding": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": 0,
                      "width": 0,
                    }
                  }
                />
              </div>
            </a>
          </div>
          <div
            className="makeStyles-followDetails-4"
          >
            <table>
              <tbody>
                <tr>
                  <td>
                    <p
                      className="MuiTypography-root MuiTypography-body1"
                    >
                      Latest release: 
                    </p>
                  </td>
                  <td>
                    <p
                      className="MuiTypography-root MuiTypography-body1"
                    >
                      Unknown
                    </p>
                  </td>
                </tr>
                <tr>
                  <td>
                    <p
                      className="MuiTypography-root MuiTypography-body1"
                    >
                      Latest chapter: 
                    </p>
                  </td>
                  <td>
                    <p
                      className="MuiTypography-root MuiTypography-body1"
                    >
                      10
                    </p>
                  </td>
                </tr>
              </tbody>
            </table>
            <ul
              className="MuiList-root makeStyles-serviceList-6 MuiList-padding"
            >
              <li
                className="MuiListItem-root makeStyles-followService-7"
                disabled={false}
              >
                <div
                  className="MuiListItemText-root makeStyles-serviceName-8"
                >
                  <span
                    className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    All services
                  </span>
                </div>
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    Unfollow
                  </span>
                </button>
              </li>
              <li
                className="MuiListItem-root makeStyles-followService-7"
                disabled={false}
              >
                <div
                  className="MuiListItemText-root makeStyles-serviceName-8"
                >
                  <span
                    className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    MangaDex
                  </span>
                </div>
                <button
                  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
                  disabled={false}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onDragLeave={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseLeave={[Function]}
                  onMouseUp={[Function]}
                  onTouchEnd={[Function]}
                  onTouchMove={[Function]}
                  onTouchStart={[Function]}
                  tabIndex={0}
                  type="button"
                >
                  <span
                    className="MuiButton-label"
                  >
                    Follow
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Follows page should render correctly should render correctly without follows 1`] = `
<div
  className="MuiContainer-root MuiContainer-disableGutters MuiContainer-maxWidthLg"
>
  <div
    className="MuiPaper-root makeStyles-root-1 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1"
    />
  </div>
</div>
`;
